src/java/main/org/apache/zookeeper/client/StaticHostProvider.java

```diff
@@ -45,6 +45,9 @@
 
     private int currentIndex = -1;
 
+    // Don't re-resolve on first next() call
+    private boolean connectedSinceNext = true;
```

Comment:
All calls to `next` and `onConnected` are from the `sendThread`. I don't see a reason for making volatile, unless we are doing it defensively. Let me know if I'm missing anything.

Note that this pull request is for the 3.4 branch, we need a different patch for 3.5 and master.